### DISCUSSION TITLE
Binding issue with Slider Value.

### DISCUSSION THREAD

**Discussion by Oen44** [AvaloniaUI/Avalonia#13812] [Q&A] [answered]
Got some binding issue with `Slider`. `TextBlock` `Text` changes perfectly whenever I move the slider or set `Value` by hand.
`CustomCanvas` `ZoomLevel` doesn't change when I move the slider, only when I hardcode `Value` on `Slider`.
Its also not setting `ZoomLevel` in my `ViewModel`. Tried using `TwoWay` mode as well, no change. What's weirder is that the initial `Value` of `Slider` is affected by initial `_zoomLevel` value in ViewModel. I did debug setter of ZoomLevel, its not executing.
Possible bug or someone notices obvious error?

View
```xml
        <ctr:CustomCanvas Grid.Row="1" ClipToBounds="True" ZoomLevel="{Binding #zoom.Value}"/>
        <Grid ColumnDefinitions="60,120" Background="#0A0A0A" Height="26" Grid.Row="2">
          <TextBlock Text="{Binding #zoom.Value, StringFormat={}{0}%}" Grid.Column="0" HorizontalAlignment="Center" VerticalAlignment="Center"></TextBlock>
          <Slider x:Name="zoom"
                  Grid.Column="1"
                  IsSnapToTickEnabled="True"
                  TickFrequency="1.0"
                  Minimum="50"
                  Maximum="200"
                  Value="{Binding ZoomLevel}"
                  VerticalAlignment="Center"/>
        </Grid>
```
ViewModel
```csharp
    private double _zoomLevel;

    public double ZoomLevel
    {
        get => _zoomLevel;
        set => this.RaiseAndSetIfChanged(ref _zoomLevel, value);
    }
```
Edit, adding CustomCanvas part
```csharp
    public static readonly StyledProperty<double> ZoomLevelProperty =
            AvaloniaProperty.Register<CustomCanvas, double>(nameof(ZoomLevel));

    public double ZoomLevel
    {
        get => GetValue(ZoomLevelProperty);
        set => SetValue(ZoomLevelProperty, value);
    }
```

**Comment by timunie** [comment]
How is ZoomLevel defined? Does it invalidate arrange and render?

**Reply by Oen44** [reply]
```csharp
    public static readonly StyledProperty<double> ZoomLevelProperty =
            AvaloniaProperty.Register<CustomCanvas, double>(nameof(ZoomLevel));

    public double ZoomLevel
    {
        get => GetValue(ZoomLevelProperty);
        set => SetValue(ZoomLevelProperty, value);
    }
```

This makes no difference.
```csharp
    public double ZoomLevel
    {
        get => GetValue(ZoomLevelProperty);
        set
        {
            SetValue(ZoomLevelProperty, value);
            InvalidateArrange();
            InvalidateVisual();
        }
    }
```

**Reply by jp2masa** [reply, marked as answer]
I think the CLR setter of a styled property doesn't have to be called to set its value. Depending on what you use the property for, you should call either `AffectsMeasure`, `AffectsArrange` or `AffectsRender` in the static constructor, something like this:

https://github.com/AvaloniaUI/Avalonia/blob/7b4bfcf7f89159b26a814cdff802af44b06116ab/src/Avalonia.Controls/StackPanel.cs#L64-L68

**Reply by Oen44** [reply]
`AffectsRender<CustomCanvas>(ZoomLevelProperty);` did the trick, thank you!
